### PR TITLE
Add recipe-k8s example: submit a Recipe job to clients in two k8s clusters

### PR DESCRIPTION
**Pull request by IsaacYangSLA**
### Description

Add `examples/advanced/recipe-k8s`, a production Recipe API example that submits one PyTorch CIFAR-10 FedAvg job through `ProdEnv` to an NVFlare system with one server and two clients running in separate Kubernetes clusters.

The example uses `set_recipe_meta` to supply:

- `JobMetaKey.RESOURCE_SPEC`: flat per-site scheduler-facing GPU requirements, which `K8sJobLauncher` also translates into `nvidia.com/gpu` requests and limits.
- `JobMetaKey.JOB_LAUNCHER_SPEC`: per-site Kubernetes job images, Python paths, CPU, memory, and ephemeral storage, plus an optional `default` block for a server prepared with `ServerK8sJobLauncher`.

The two clients train a bundled PyTorch model on disjoint CIFAR-10 partitions. The example also documents shared and per-site workload images, CPU/GPU behavior, expected output, troubleshooting, and the division of responsibilities between data scientists and platform teams.

Also add the example to the examples index, link it from the Recipe Metadata guide, document submitting it from the Brev scripted three-cluster quickstart, and verify torchvision in the maintained workload image.

Verified by exporting the job and inspecting the generated `meta.json` and app packages, and by running 119 targeted tests plus scoped Black, isort, and flake8 checks.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [x] Documentation updated.

### CODE REVIEW

File: examples/advanced/recipe-k8s/job.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+from nvflare.apis.job_def import JobMetaKey
+from nvflare.app_common.np.recipes.fedavg import NumpyFedAvgRecipe
+from nvflare.recipe import ProdEnv, set_recipe_meta
+
+
+def non_negative_int(value: str) -> int:
+    value = int(value)
+    if value < 0:
+        raise argparse.ArgumentTypeError("must be greater than or equal to 0")
+    return value
```
[the remark this answers is not in the document]
Comment (IsaacYangSLA) [reply examples/advanced/recipe-k8s/job.py]: Addressed in 4f1dcfbf0. `non_negative_int` now preserves the raw string, catches `ValueError`, raises `argparse.ArgumentTypeError` with the invalid value, and returns the parsed integer. I also verified the invalid-value CLI path.

File: examples/advanced/recipe-k8s/job.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        # their cluster-specific images and settings.
+        launcher_spec["default"] = {
+            "k8s": k8s_launcher_spec(
+                args.server_image,
+                "1",
+                "2Gi",
+                args.python_path,
+                args.ephemeral_storage,
+            )
+        }
```
[the remark this answers is not in the document]
Comment (IsaacYangSLA) [reply examples/advanced/recipe-k8s/job.py:247]: Addressed in 4f1dcfbf0. The example now provides `--server-cpu` and `--server-memory` (defaults `1` and `2Gi`), wires them into the default Kubernetes launcher spec, and documents them. I verified an export with `--server-cpu 4 --server-memory 8Gi` contains those values.

File: examples/advanced/recipe-k8s/job.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    parser.add_argument("--python-path", default="/usr/local/bin/python3", help="Python executable in the job images.")
+    parser.add_argument("--ephemeral-storage", default="2Gi", help="Workspace storage for each Kubernetes job pod.")
+    parser.add_argument(
+        "--server-image",
+        help=(
+            "Job image for a server that also uses a Kubernetes job launcher. "
+            "Omit this when the server uses the process launcher."
+        ),
+    )
+    return parser
```
[the remark this answers is not in the document]
Comment (IsaacYangSLA) [reply examples/advanced/recipe-k8s/job.py:119]: These flags are owned and consumed by the Recipe API before this parser processes example-specific options, so defining them again here would duplicate framework behavior. I verified the documented command exports successfully. Commit 4f1dcfbf0 clarifies this in both `--help` and the README.

File: examples/advanced/recipe-k8s/README.md [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+  --site-2-memory 32Gi
+```
+
+If the server is also prepared with `ServerK8sJobLauncher`, add an image that
+its cluster can pull:
+
+```bash
+  --server-image registry-server.example.com/nvflare-job:2.9
+```
+
```
[the remark this answers is not in the document]
Comment (IsaacYangSLA) [reply examples/advanced/recipe-k8s/README.md:169]: These are standard Recipe API flags consumed by `nvflare.recipe` before this example's `argparse` parser runs. `Recipe.execute()` then exports and exits without invoking `ProdEnv.deploy()`. I verified the documented command exports successfully without connecting. Commit 4f1dcfbf0 adds an argparse epilog and README clarification instead of duplicating framework-owned flags.

File: examples/advanced/recipe-k8s/job.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        username=args.username,
+        study=args.study,
+    )
+
+    run = recipe.execute(env)
+    print(f"Job ID: {run.get_job_id()}")
+    result = run.get_result()
+    print(f"Job status: {run.get_status()}")
+    print(f"Result: {result}")
+
```
[the remark this answers is not in the document]
Comment (IsaacYangSLA) [reply examples/advanced/recipe-k8s/job.py:270]: The gate already exists in `Recipe.execute()`: after the Recipe API consumes `--export`/`--export-dir`, it calls `export()` and exits before `ProdEnv.deploy()`. I verified the documented command generates the job config without connecting to a server. Commit 4f1dcfbf0 documents that behavior explicitly.

**Review by chesterxgchen — COMMENTED** [review]
should we use a bit more realistic example, CIFAR10 or something else ?

**Comment by IsaacYangSLA** [comment]
> should we use a bit more realistic example, CIFAR10 or something else ?

Let me change it.  My original thought was the actual training didn't matter since we just tried to emphasize the key points.

**Review by chesterxgchen — COMMENTED** [review]
from claude
User-experience review (data scientist running this)

1. README paragraph order bug ("Inspect Without Submitting"): the sentence "…because ProdEnv validates it during construction:" ends with a colon promising a code block, but the --export/define_parser() explanation paragraph intervenes before the block appears. Move that paragraph above the intro sentence (or after the code block). This is the one change I'd definitely request.

2. No "expected output" or troubleshooting section. For a k8s example the failure surface is large (image pull errors, RBAC, mismatched site names, admin cert issues). A short "you should see" block (job ID printed, kubectl get pods showing the job pod, final status) plus 2–3 common-failure hints would be the highest-value addition for a first-time user.

3. Silent blocking wait. After Job ID: … prints, run.get_result() blocks with no output until the job finishes. A print("Waiting for job to complete…") before it — or a README pointer to watch pods with kubectl — would prevent the "is it hung?" moment.

4. Validation errors surface as tracebacks. create_recipe raises ValueError for same site names / default / --num-rounds < 1, which a data scientist sees as a crash. Prefer parser.error(...) (clean two-line exit) or a positive_int argparse type mirroring the existing non_negative_int.

5. Minor friction: the common case of one registry serving both clusters still requires typing the image twice. An optional --image shorthand (used when the per-site flags are omitted) would shorten the happy path. Optional.

Chief-data-scientist review (is this the right story to show a team)

- The separation-of-concerns narrative is the strongest part and it's told correctly: participant identity vs. cluster, resource_spec (what the scheduler needs) vs. launcher_spec (what the container needs), and the explicit "no kubeconfig/namespace/endpoint belongs in job metadata" statement. That's exactly the mental model needed to delegate safely between DS and infra teams.

- One framing gap: in practice, image URLs, CPU/memory, and python paths come from the platform team, not the data scientist. A one-sentence note that a team would typically wrap job.py with org defaults (or receive these values from ops) would complete the role-separation story.

- The BYOC-authorization prerequisite and the "training script is bundled, not baked into the image" point are both called out — good, those are the two things a CDS asks about first.

**Comment by IsaacYangSLA** [comment]
Addressed the new review feedback in 9e015486a.

- Replaced the toy NumPy update with a PyTorch CIFAR-10 FedAvg example using disjoint per-site partitions, bounded demo sample counts, explicit CPU/GPU intent, and bundled client/model code.
- Added a shared `--image` fallback with per-site overrides, clean argparse validation, and a visible waiting message.
- Fixed the export-section paragraph order and added expected output, Kubernetes troubleshooting, platform-team ownership guidance, and separate parent/workload image instructions.
- Added focused coverage for CLI errors, Recipe export metadata, code bundling, image precedence, partitioning, model training/evaluation, and device selection.

Validation: 117 targeted tests passed; scoped Black/isort/flake8 and `git diff --check` passed; the documented `--export` flow was verified end to end.

**Comment by chesterxgchen** [comment]
Remaining nits (non-blocking)

- Stale PR description: it still says "submits one NumPy FedAvg job" — the example is now PyTorch CIFAR-10. Worth a one-line edit before merge.
- --job-name feeds the export directory name but isn't validated (whitespace/slashes), while --data-dir is. Harmless since it's not embedded in task_script_args; mention only for symmetry.

File: examples/advanced/recipe-k8s/job.py [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+        # their cluster-specific images and settings.
+        launcher_spec["default"] = {
+            "k8s": k8s_launcher_spec(
+                args.server_image,
+                args.server_cpu,
+                args.server_memory,
+                args.python_path,
+                args.ephemeral_storage,
+            )
+        }
```
Comment (YuanTingHsieh) [inline examples/advanced/recipe-k8s/job.py:247]: for the current code we need this, same as DockerJobLauncher, we might need the feature that

study_job.json at each site has study name map to image name. 

So when people submit job under this study, they dont need to specify image anymore.

**Review by YuanTingHsieh — DISMISSED** [review]

**Comment by IsaacYangSLA** [comment]
Addressed both remaining nits in 983af1a9b: the PR description now reflects the PyTorch CIFAR-10 example and current validation, and `--job-name` now uses NVFlare's canonical job-name format validation to reject whitespace and path separators cleanly. Added regression coverage for both invalid forms; 119 targeted tests and scoped Black/isort/flake8 checks pass.

**Review by YuanTingHsieh — APPROVED** [review]
